File: packages/core/src/strategies.ts
import type { Provider, Model } from './types';

export interface ProviderStrategy {
    format: string;
    // 模型列表
    getModelsEndpoint?: (baseUrl: string, apiKey: string) => string;
    parseModelsResponse?: (data: any) => Model[];
    // 聊天 (也用于连接测试)
    getChatEndpoint: (baseUrl: string, apiKey: string, model: string) => string;
    buildChatPayload: (model: string, messages: Array<{ role: string; content: string }>, maxTokens?: number) => Record<string, unknown>;
    parseChatResponse: (data: any) => string;
    // Headers
    buildHeaders: (apiKey: string) => Record<string, string>;
}

export const defaultParseModelsResponse = (data: any): Model[] => {
    if (Array.isArray(data?.data)) {
        return data.data
            .filter((m: any) => m.id)
            .map((m: any) => {
                const name =
                    m.name || m.display_name ||
                    (m.id.split('/').pop() ?? '')
                        .replace(/[-_]/g, ' ')
                        .replace(/\b\w/g, (c: string) => c.toUpperCase());
                return {
                    id: m.id,
                    name,
                    created: m.created || 0,
                } as any;
            })
            .sort((a: any, b: any) => {
                const diff = (b.created || 0) - (a.created || 0);
                if (diff !== 0) return diff;
                return (b.id || '').localeCompare(a.id || '');
            }); // 最新的排在前面
    }
    return [];
};

const openaiStrategy: ProviderStrategy = {
    format: 'openai',
    getModelsEndpoint: (baseUrl) => `${baseUrl}/models`,
    getChatEndpoint: (baseUrl) => `${baseUrl}/chat/completions`,
    buildHeaders: (apiKey) => {
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${apiKey}`,
        };
        // OpenRouter compatibility & Best Practices
        if (typeof window !== 'undefined' && window.location) {
            headers['HTTP-Referer'] = window.location.origin;
            headers['X-Title'] = document.title || 'AI Selector';
        }
        return headers;
    },
    buildChatPayload: (model, messages, maxTokens) => {
        const payload: any = {
            model,
            messages,
        };
        if (maxTokens) {
            payload.max_completion_tokens = maxTokens;
        }
        return payload;
    },
    parseChatResponse: (data) => {
        return data.choices?.[0]?.message?.content || '';
    },
};

const anthropicStrategy: ProviderStrategy = {
    format: 'anthropic',
    getModelsEndpoint: (baseUrl) => `${baseUrl}/models`,
    getChatEndpoint: (baseUrl) => `${baseUrl}/messages`,
    buildHeaders: (apiKey) => ({
        'Content-Type': 'application/json',
        'x-api-key': apiKey,
        'anthropic-version': '2023-06-01',
        'anthropic-dangerous-direct-browser-access': 'true',  // 启用浏览器直连 CORS
    }),
    buildChatPayload: (model, messages, maxTokens) => {
        const payload: any = {
            model,
            messages,
            max_tokens: maxTokens || 1024,  // Anthropic API 强制要求 max_tokens
        };
        return payload;
    },
    parseChatResponse: (data) => {
        return data.content?.[0]?.text || '';
    },
    // Anthropic 返回: { data: [{ id, display_name, created_at }] }
    parseModelsResponse: (data) => {
        if (Array.isArray(data.data)) {
            return data.data
                .filter((m: any) => m.id)
                .map((m: any) => ({
                    id: m.id,
                    name: m.display_name || m.id,
                    created: m.created_at ? new Date(m.created_at).getTime() / 1000 : 0,
                }))
                .sort((a: any, b: any) => (b.created || 0) - (a.created || 0));
        }
        return [];
    },
};

const geminiStrategy: ProviderStrategy = {
    format: 'gemini',
    getModelsEndpoint: (baseUrl, apiKey) => `${baseUrl}/models?key=${apiKey}`,
    getChatEndpoint: (baseUrl, apiKey, model) => `${baseUrl}/models/${model}:generateContent?key=${apiKey}`,
    buildHeaders: () => ({
        'Content-Type': 'application/json',
    }),
    buildChatPayload: (_model, messages, maxTokens) => {
        // 转换 OpenAI 格式的 messages 为 Gemini 格式
        const contents = messages.map(m => ({
            role: m.role === 'assistant' ? 'model' : 'user',
            parts: [{ text: m.content }],
        }));
        const payload: any = { contents };
        if (maxTokens) {
            payload.generationConfig = { maxOutputTokens: maxTokens };
        }
        return payload;
    },
    parseChatResponse: (data) => {
        return data.candidates?.[0]?.content?.parts?.[0]?.text || '';
    },
    // Gemini 返回格式: { models: [{ name: "models/gemini-pro", ... }] }
    parseModelsResponse: (data) => {
        if (Array.isArray(data.models)) {
            return data.models
                .filter((m: any) => m.supportedGenerationMethods?.includes('generateContent'))
                .map((m: any) => ({
                    id: m.name.replace('models/', ''), // "models/gemini-pro" -> "gemini-pro"
                    name: m.displayName || m.name.replace('models/', ''),
                    created: m.created || 0,
                }))
                .sort((a: any, b: any) => (b.id || '').localeCompare(a.id || ''));
        }
        return [];
    },
};

const cohereStrategy: ProviderStrategy = {
    format: 'cohere',
    getModelsEndpoint: (baseUrl) => `${baseUrl}/models`,
    getChatEndpoint: (baseUrl) => `${baseUrl}/chat`,
    buildHeaders: (apiKey) => ({
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`,
    }),
    buildChatPayload: (model, messages, maxTokens) => {
        // Cohere v2 API 使用 messages 格式
        const payload: any = {
            model,
            messages,
        };
        if (maxTokens) {
            payload.max_tokens = maxTokens;
        }
        return payload;
    },
    parseChatResponse: (data) => {
        // Cohere v2 响应格式
        return data.message?.content?.[0]?.text || '';
    },
};

export const strategyRegistry: Record<string, ProviderStrategy> = {
    openai: openaiStrategy,
    anthropic: anthropicStrategy,
    gemini: geminiStrategy,
    cohere: cohereStrategy,
};

export function getStrategy(format: string): ProviderStrategy {
    return strategyRegistry[format] || openaiStrategy;
}

// ============================================================================
// 纯前端直连聊天函数
// ============================================================================

export interface DirectChatOptions {
    apiFormat: string;
    baseUrl: string;
    apiKey: string;
    model: string;
    messages: Array<{ role: string; content: string }>;
    maxTokens?: number;
}

export interface DirectChatResult {
    success: boolean;
    content?: string;
    message?: string;
    latencyMs?: number;
}

/**
 * 纯前端直连 AI 厂商进行聊天
 * 注意: 这会将 API Key 暴露在浏览器中，仅适用于 Demo/测试场景
 */
export async function sendDirectChat(options: DirectChatOptions): Promise<DirectChatResult> {
    const { apiFormat, baseUrl, apiKey, model, messages, maxTokens } = options;
    const strategy = getStrategy(apiFormat);

    const endpoint = strategy.getChatEndpoint(baseUrl, apiKey, model);
    const headers = strategy.buildHeaders(apiKey);
    const payload = strategy.buildChatPayload(model, messages, maxTokens);

    const startTime = performance.now();

    try {
        const response = await fetch(endpoint, {
            method: 'POST',
            headers,
            body: JSON.stringify(payload),
        });

        const latencyMs = Math.round(performance.now() - startTime);

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            return {
                success: false,
                message: errorData.error?.message || `HTTP ${response.status}: ${response.statusText}`,
                latencyMs,
            };
        }

        const data = await response.json();
        const content = strategy.parseChatResponse(data);

        return {
            success: true,
            content,
            latencyMs,
        };
    } catch (e) {
        return {
            success: false,
            message: e instanceof Error ? e.message : '网络错误',
        };
    }
}

// ============================================================================
// 连接测试函数 (复用聊天接口)
// ============================================================================

export interface TestConnectionOptions {
    apiFormat: string;
    baseUrl: string;
    apiKey: string;
    model: string;
}

export interface TestConnectionResult {
    success: boolean;
    latencyMs?: number;
    message?: string;
}

/**
 * 测试 AI 厂商连接 (通过发送一个简单的聊天请求)
 */
export async function testDirectConnection(options: TestConnectionOptions): Promise<TestConnectionResult> {
    const result = await sendDirectChat({
        apiFormat: options.apiFormat,
        baseUrl: options.baseUrl,
        apiKey: options.apiKey,
        model: options.model,
        messages: [{ role: 'user', content: 'Hi' }],
        // maxTokens: 5, // 不设置 maxTokens 以兼容 o1 等不支持该参数的模型
    });

    return {
        success: result.success,
        latencyMs: result.latencyMs,
        message: result.success ? undefined : result.message,
    };
}

File: backend/strategies.py
"""
聊天策略模式实现

每个 API 格式对应一个策略类，负责构建请求和解析响应。
"""

from abc import ABC, abstractmethod
from typing import Dict, Any, List, Type, Optional, Tuple
import httpx

from models import ChatResult


# ============================================================================
# 抽象基类
# ============================================================================

class ChatStrategy(ABC):
    """聊天策略抽象基类"""
    
    # 是否支持 /models 端点
    supports_models_api: bool = True
    
    @abstractmethod
    def build_headers(self, api_key: str) -> Dict[str, str]:
        """构建聊天请求头"""
        pass
    
    @abstractmethod
    def build_endpoint(self, base_url: str, model: str, api_key: str) -> str:
        """构建聊天请求 URL"""
        pass
    
    @abstractmethod
    def build_payload(self, model: str, messages: List[Dict[str, str]], max_tokens: int) -> Dict[str, Any]:
        """构建聊天请求体"""
        pass
    
    @abstractmethod
    def parse_response(self, data: Dict[str, Any], model: str) -> ChatResult:
        """解析聊天响应"""
        pass
    
    def build_models_request(self, base_url: str, api_key: str) -> Tuple[str, Dict[str, str]]:
        """
        构建获取模型列表的请求
        
        Returns:
            (endpoint, headers) 元组
        """
        # 默认实现：OpenAI 兼容格式
        return (
            f"{base_url}/models",
            {"Authorization": f"Bearer {api_key}"} if api_key else {}
        )
    
    def parse_models_response(self, data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        解析模型列表响应
        
        Returns:
            [{"id": "...", "name": "...", "created": ...}, ...]
        """
        models = []
        for m in data.get("data", []):
            model_id = m.get("id")
            if model_id:
                # 优先取 name，取不到则将 id 格式化作为 name
                name = m.get("name") or model_id.split("/")[-1].replace("-", " ").replace("_", " ").title()
                models.append({
                    "id": model_id,
                    "name": name,
                    "created": m.get("created", 0)  # 保留创建时间用于排序
                })
        return models
    
    async def execute(
        self,
        client: httpx.AsyncClient,
        base_url: str,
        api_key: str,
        model: str,
        messages: List[Dict[str, str]],
        max_tokens: int,
    ) -> ChatResult:
        """执行聊天请求 (模板方法)"""
        headers = self.build_headers(api_key)
        endpoint = self.build_endpoint(base_url, model, api_key)
        payload = self.build_payload(model, messages, max_tokens)
        
        response = await client.post(endpoint, headers=headers, json=payload)
        response.raise_for_status()
        return self.parse_response(response.json(), model)
    
    async def fetch_models(
        self,
        client: httpx.AsyncClient,
        base_url: str,
        api_key: str,
    ) -> List[Dict[str, str]]:
        """获取模型列表"""
        if not self.supports_models_api:
            return []
        
        endpoint, headers = self.build_models_request(base_url, api_key)
        response = await client.get(endpoint, headers=headers)
        response.raise_for_status()
        return self.parse_models_response(response.json())


# ============================================================================
# 策略实现
# ============================================================================

class OpenAIStrategy(ChatStrategy):
    """OpenAI 兼容格式 (DeepSeek, Groq, Moonshot, etc.)"""
    
    def build_headers(self, api_key: str) -> Dict[str, str]:
        return {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    
    def build_endpoint(self, base_url: str, model: str, api_key: str) -> str:
        return f"{base_url}/chat/completions"
    
    def build_payload(self, model: str, messages: List[Dict[str, str]], max_tokens: int) -> Dict[str, Any]:
        return {"model": model, "messages": messages, "max_tokens": max_tokens}
    
    def parse_response(self, data: Dict[str, Any], model: str) -> ChatResult:
        content = data.get("choices", [{}])[0].get("message", {}).get("content", "") if data.get("choices") else ""
        usage = data.get("usage", {})
        return ChatResult(
            raw_response=data, content=content, model=data.get("model", model),
            usage={"prompt_tokens": usage.get("prompt_tokens", 0), "completion_tokens": usage.get("completion_tokens", 0), "total_tokens": usage.get("total_tokens", 0)}
        )


class AnthropicStrategy(ChatStrategy):
    """Anthropic Claude 格式"""
    
    # Anthropic 不提供 /models 端点
    supports_models_api = False
    
    def build_headers(self, api_key: str) -> Dict[str, str]:
        return {"x-api-key": api_key, "anthropic-version": "2023-06-01", "Content-Type": "application/json"}
    
    def build_endpoint(self, base_url: str, model: str, api_key: str) -> str:
        return f"{base_url}/messages"
    
    def build_payload(self, model: str, messages: List[Dict[str, str]], max_tokens: int) -> Dict[str, Any]:
        return {"model": model, "messages": messages, "max_tokens": max_tokens}
    
    def parse_response(self, data: Dict[str, Any], model: str) -> ChatResult:
        content = "".join(b.get("text", "") for b in data.get("content", []) if b.get("type") == "text")
        usage = data.get("usage", {})
        inp, out = usage.get("input_tokens", 0), usage.get("output_tokens", 0)
        return ChatResult(
            raw_response=data, content=content, model=data.get("model", model),
            usage={"prompt_tokens": inp, "completion_tokens": out, "total_tokens": inp + out}
        )


class GeminiStrategy(ChatStrategy):
    """Google Gemini 格式"""
    
    def build_headers(self, api_key: str) -> Dict[str, str]:
        return {"Content-Type": "application/json"}
    
    def build_endpoint(self, base_url: str, model: str, api_key: str) -> str:
        return f"{base_url}/models/{model}:generateContent?key={api_key}"
    
    def build_payload(self, model: str, messages: List[Dict[str, str]], max_tokens: int) -> Dict[str, Any]:
        contents = [{"role": "user" if m.get("role") == "user" else "model", "parts": [{"text": m.get("content", "")}]} for m in messages]
        return {"contents": contents, "generationConfig": {"maxOutputTokens": max_tokens}}
    
    def parse_response(self, data: Dict[str, Any], model: str) -> ChatResult:
        content = "".join(p.get("text", "") for p in data.get("candidates", [{}])[0].get("content", {}).get("parts", []))
        meta = data.get("usageMetadata", {})
        return ChatResult(
            raw_response=data, content=content, model=model,
            usage={"prompt_tokens": meta.get("promptTokenCount", 0), "completion_tokens": meta.get("candidatesTokenCount", 0), "total_tokens": meta.get("totalTokenCount", 0)}
        )
    
    def build_models_request(self, base_url: str, api_key: str) -> Tuple[str, Dict[str, str]]:
        # Gemini 使用 URL 参数鉴权
        return (f"{base_url}/models?key={api_key}", {})
    


class CohereStrategy(ChatStrategy):
    """Cohere v2 格式"""
    
    def build_headers(self, api_key: str) -> Dict[str, str]:
        return {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    
    def build_endpoint(self, base_url: str, model: str, api_key: str) -> str:
        return f"{base_url}/chat"
    
    def build_payload(self, model: str, messages: List[Dict[str, str]], max_tokens: int) -> Dict[str, Any]:
        return {"model": model, "messages": messages, "max_tokens": max_tokens}
    
    def parse_response(self, data: Dict[str, Any], model: str) -> ChatResult:
        content = "".join(b.get("text", "") for b in data.get("message", {}).get("content", []))
        usage = data.get("usage", {}).get("billed_units", {})
        inp, out = usage.get("input_tokens", 0), usage.get("output_tokens", 0)
        return ChatResult(
            raw_response=data, content=content, model=model,
            usage={"prompt_tokens": inp, "completion_tokens": out, "total_tokens": inp + out}
        )
    


# ============================================================================
# 策略注册表
# ============================================================================

STRATEGY_REGISTRY: Dict[str, Type[ChatStrategy]] = {
    "openai": OpenAIStrategy,
    "anthropic": AnthropicStrategy,
    "gemini": GeminiStrategy,
    "cohere": CohereStrategy,
}


def get_strategy(api_format: str) -> ChatStrategy:
    """根据 API 格式获取策略实例"""
    cls = STRATEGY_REGISTRY.get(api_format)
    if not cls:
        raise ValueError(f"不支持的 API 格式: {api_format}。支持: {', '.join(STRATEGY_REGISTRY.keys())}")
    return cls()
